inst/training_202307/202307.qmd

```diff
@@ -0,0 +1,448 @@
+
+:::{.callout-important}
+You have to specify minimal versions allowing dependency resolution to complete.
+:::
+
+It could be a process of multiple iterations of increasing dependent package versions.
+
+## Example - failure
+
+:::: {.columns}
+
+::: {.column width="50%"}
+```{.yaml code-line-numbers="7-8"}
+Type: Package
+Package: foo.package
+Version: 1.2.3.9000
+Depends:
+    R (>= 3.6)
+Imports:
+    dplyr (>= 1.0.0),
+    tibble (>= 1.0.0)
+    (...)
+```
+:::
+
+::: {.column width="50%"}
+```{.yaml code-line-numbers="2,3,8"}
+Type: Package
+Package: dplyr
+Package: 1.0.0
+Depends:
+    R (>= 3.2.0)
+Imports:
+    (...)
+    tibble (>= 2.1.3)
+    (...)
+```
+:::
+
+::::
+
+<br>
+
+```{r}
+#| eval: false
+#| echo: true
+#| code-line-numbers: false
+x <- pkgdepends::new_pkg_deps(c("dplyr@1.0.0", "tibble@1.0.0"))
+x$solve()
+x$get_solution()
+```
+
+```.default
+<pkg_solution>
++ result: FAILED
++ refs:
+  - dplyr@1.0.0
+  - tibble@1.0.0
++ constraints (93):
+  (...)
+x failures:
+* dplyr@1.0.0: Can't install dependency tibble (>= 2.1.3)
+* tibble@1.0.0: Conflicts with tibble@1.0.0
+* tibble: Conflicts with tibble@1.0.0
+```
+
+## Example - success
+
+:::: {.columns}
+
+::: {.column width="50%"}
+```{.yaml code-line-numbers="8"}
+Type: Package
+Package: foo.package
+Version: 1.2.3.9000
+Depends:
+    R (>= 3.6)
+Imports:
+    dplyr (>= 1.0.0),
+    tibble (>= 2.1.3)
+    (...)
+```
+:::
+
+::: {.column width="50%"}
+```{.yaml code-line-numbers="8"}
+Type: Package
+Package: dplyr
+Package: 1.0.0
+Depends:
+    R (>= 3.2.0)
+Imports:
+    (...)
+    tibble (>= 2.1.3)
+    (...)
+```
+:::
+
+::::
+
+<br>
+
+```{r}
+#| eval: false
+#| echo: true
+#| code-line-numbers: false
+x <- pkgdepends::new_pkg_deps(c("dplyr@1.0.0", "tibble@2.1.3"))
+x$solve()
+x$get_solution()
+```
+
+```.default
+<pkg_solution>
++ result: OK
++ refs:
+  - dplyr@1.0.0
+  - tibble@2.1.3
++ constraints (95):
+  - select dplyr exactly once
+  - select tibble exactly once
+  - select R6 at most once
+  - select cli at most once
+  - select ellipsis at most once
+  - select fansi at most once
+  - select generics at most once
+  - select glue at most once
+  - select lifecycle at most once
+  - select magrittr at most once
+  ...
++ solution:
+  - cli
+  - crayon
+  - dplyr@1.0.0
+  - ellipsis
+  - fansi
+  - generics
+  - glue
+  - lifecycle
+  - magrittr
+  - pillar
+  - pkgconfig
+  - R6
+  - rlang
+  - tibble@2.1.3
+  - tidyselect
+  - utf8
+  - vctrs
+  - withr
+```
+
+## Discussion:
+
+- One might argue that `foo.package` is fully compatible with the `tibble@1.0.0` syntax but the truth is that `tibble@1.0.0` cannot be installed alongside `dplyr@1.0.0` hence `foo.package` cannot be used with both `dplyr@1.0.0` and `tibble@1.0.0` loaded. Therefore `tibble (>= 1.0.0)` really indicates `tibble (>= 2.1.3)` in this case.
+
+- Because of above, the minimal version specification of dependencies is the _true_ **installable** and **compatible** set of dependencies.
```

Comment:
Disagree here. Conceptually, `foo.package`  only needs tibble > 1.0.0 and dplyr 1.0.0. It does not care that dplyr 1.0.0 requires tibble 2.1.3. Thats dplyr's problem, and that requirement is already encapsulated and enforced in dplyr 1.0.0

Look at it this way. The point of this is to figure out if our packge is usable within a particular renv'ed environment, right. that renv has tibble 1.0.0, it isn't going to have dplyr 1.0.0, so we already know the answer is no. We don't get any extra information from also requiring tibble 2.1.3

Don't get me wrong, I'm a big fan of installable cohorts. I have literally written a paper and given numerous research talks about it. It was central in my promotion talk to scientist at genentech. _**BUT**_, that isn't the purpose of htis tool. No one is going to come along and say "man I really want to install the exact minimum requirements to use rlistings 0.5.3, _but no later because I hate all package versions that have come out in the last 5 years!!!_

This tool is for finding requirements, and in that case being discussed here, the only meaningful versioned requirement is dplyr 1.0.0. So long as you ahve that, you're guaranteed to have a version of tibble that works too, so you're good and nothing more need be stated.

Reply:
I was expecting this comment :) It is probably the weakest side and encapsulation and who is responsible of what resonates with me very well.

It all stems from the limitations of `pkgdepends` and its resolve dependency algorithm - or to be more fair: our specific requirements.

Please have a look at the example up above - let me re-paste it again here:
```{r}
x <- pkgdepends::new_pkg_deps(c("dplyr@1.0.0", "tibble@1.0.0"))
x$solve()
x$get_solution()
```

```.default
<pkg_solution>
+ result: FAILED
+ refs:
  - dplyr@1.0.0
  - tibble@1.0.0
+ constraints (93):
  (...)
x failures:
* dplyr@1.0.0: Can't install dependency tibble (>= 2.1.3)
* tibble@1.0.0: Conflicts with tibble@1.0.0
* tibble: Conflicts with tibble@1.0.0
```

This pair of packages are not installable together as there is an obvious conflict that `tibble@1.0.0` does not meet version criteria within `dplyr`. It's a significant limitation that we have to bear in mind if we want to keep it as a thin layer on `pkgdepends`.

I have spent considerable amount of time of how to overcome this (and also fit nicely within the whole framework) with no success :( I can't really ask package maintainers for it as I am aware that this is a very specific requirement. This is true that currently named "min" deps is not a code-compabilitywise set of minimal dependencies - it's more of an installable (not to mention compileable) set of packages that is a bit more strict. I will emphasise this once more in the slides.
My view on this is that even if it's imperfect - this is a big step forward. I totally got your point about no-one asking for a specific very old version of a package. This is very unlikely to happen. I was thinking more about set of interdependent packages that are developed in parallel - something we are doing a lot. Always blindly vbumping deps is probably not the best strategy. On the other hand - keeping it unchanged is not correct either. Wanted to find a good tool that will warn you when vbump is needed.

Reply:
This is... somewhat true, but backwards. The cascade goes the other way

1. You cannot load your hard dependency (dplyr needs tibble >=2.1.3 and can't get it), so
2. you cannot load your package, because you must load all hard dependencies to load your package
3. you cannot install your package (because the installation process includes loading the package, which would fail even if you slipped past the other checks)

Note how I'm framing this. It has nothing to do with tibble per se. Its the dplyr requirement for our package that failed here. That's important. 

Packages (which have non-base-R dependencies) can **_never_** be considered in isolation. Ever. For anything, in any circumstances. 

But this is a deeper issue. There was no reason to force tibble 2.0.0 here when one of the active dependencies forced a higher version. The fact that that happened is a sign something in the algorithm isn't doing the right thing.

Put another way, if dplyr 1.0.0 requires 2.1.3 (and has declared as much), there was no reason to even try with tibble 2.0.0 and dplyr 1.0.0. It was literally impossible for that test to be successful a priori. Thus it was impossible for us to get any new information from running it.

Reply:
> But this is a deeper issue. There was no reason to force tibble 2.0.0 here when one of the active dependencies forced a higher version. The fact that that happened is a sign something in the algorithm isn't doing the right thing.

It was just an experiment (not part of any algorihtm) of an environment that reflects 1:1 with what specified by direct dependencies. 

> Put another way, if dplyr 1.0.0 requires 2.1.3 (and has declared as much), there was no reason to even try with tibble 2.0.0 and dplyr 1.0.0. It was literally impossible for that test to be successful a priori. Thus it was impossible for us to get any new information from running it.

Yes exactly. That was the goal for it - to show that such environment is non-fuctional and malformed. This makes me think that such configuration of dependencies is not really testable hence: not correct? (don't know if "correct" is the right term here but I hope you got my point). You can't specify `tibble@2.0.0` + `dplyr@1.0.0` as you are not able to validate its correctness.

Like I said - some adjustment (here: vbump tibble) needs to be done. In the current implementation it just errors. This is well embedded in the dependency resolution algorighm in pkgdepends. I am currently playing with my own resolution logic that uses min for direct and concurrent for indirect deps. It's definitely not super nice as I am relaying on private class fields but it's sort of working. Let me polish it and I will reach out to you again soon for feedback. But to be clear: that not means that I am convinced about correctness of dependencies specifications. It's just a way how to handle incorrect inputs and correct it on-the-fly.